File: backend/docs/api/merge.js
const fs = require("fs");
let b = fs.readFileSync("phone-store-openapi.yaml", "utf8");

const duplicates = [
  "/api/v1/warehouses/{id}",
  "/api/v1/suppliers/{id}",
  "/api/v1/orders/{orderCode}/return-requests",
  "/api/v1/admin/refunds",
];

duplicates.forEach((path) => {
  // Regex to match exact path declaration e.g. "  /api/v1/warehouses/{id}:\n"
  const safePath = path
    .replace(/\\/ / g, "\\/")
    .replace(/\\{/g, "\\{")
    .replace(/\\}/g, "\\}");
  const regex = new RegExp("\\n  " + safePath + ":\\n", "g");

  const parts = b.split(regex);
  if (parts.length === 3) {
    // The first part is everything before the FIRST occurrence
    // The second part is everything between FIRST and SECOND occurrence
    // The third part is everything after the SECOND occurrence

    // We want to extract the methods from the second occurrence (parts[2])
    const endOfBottomBlock = parts[2].search(/\n  \/[a-zA-Z]/);
    let bottomChildren = "";

    if (endOfBottomBlock === -1) {
      // It's the very last path in the file
      bottomChildren = parts[2].substring(0, parts[2].search(/\ncomponents:/));
      if (bottomChildren === "") {
        bottomChildren = parts[2];
        parts[2] = "";
      } else {
        parts[2] = parts[2].substring(bottomChildren.length);
      }
    } else {
      bottomChildren = parts[2].substring(0, endOfBottomBlock);
      parts[2] = parts[2].substring(endOfBottomBlock);
    }

    // Move the bottom children to right after the first occurrence
    b =
      parts[0] + "\\n  " + path + ":\\n" + bottomChildren + parts[1] + parts[2];
  }
});

fs.writeFileSync("phone-store-openapi.yaml", b);
console.log("Merge complete!");

File: backend/docs/api/merge.py
import re

with open('phone-store-openapi.yaml', 'r', encoding='utf-8') as f:
    text = f.read()

dups = [
    '/api/v1/warehouses/{id}',
    '/api/v1/suppliers/{id}',
    '/api/v1/orders/{orderCode}/return-requests',
    '/api/v1/admin/refunds'
]

for dup in dups:
    pattern = r'\n  ' + re.escape(dup) + r':\n'
    parts = re.split(pattern, text)
    if len(parts) == 3:
        top = parts[0]
        middle = parts[1]
        bottom = parts[2]
        
        # Bottom block goes from start of bottom up to the next route or end of components
        next_route_match = re.search(r'\n  /[a-zA-Z]', bottom)
        if next_route_match:
            end_idx = next_route_match.start()
            bottom_block = bottom[:end_idx]
            rest_bottom = bottom[end_idx:]
        else:
            comp_match = re.search(r'\ncomponents:', bottom)
            if comp_match:
                end_idx = comp_match.start()
                bottom_block = bottom[:end_idx]
                rest_bottom = bottom[end_idx:]
            else:
                bottom_block = bottom
                rest_bottom = ''
                
        # Merge bottom_block into the first occurrence
        text = top + '\n  ' + dup + ':\n' + bottom_block.strip('\n') + '\n' + middle + rest_bottom

with open('phone-store-openapi.yaml', 'w', encoding='utf-8') as f:
    f.write(text)

print("Merged perfectly via Python!")
